instrumentation/apache-camel-2.20/src/main/java/io/opentelemetry/javaagent/instrumentation/apachecamel/CamelEventNotifier.java

```diff
@@ -0,0 +1,95 @@
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * <p>http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * <p>Unless required by applicable law or agreed to in writing, software distributed under the
+ * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.javaagent.instrumentation.apachecamel;
+
+import io.grpc.Context;
+import io.opentelemetry.trace.Span;
+import java.util.EventObject;
+import org.apache.camel.management.event.ExchangeSendingEvent;
+import org.apache.camel.management.event.ExchangeSentEvent;
+import org.apache.camel.support.EventNotifierSupport;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+final class CamelEventNotifier extends EventNotifierSupport {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CamelEventNotifier.class);
+
+  @Override
+  public void notify(EventObject event) {
+
+    try {
+      /** Camel about to send (outbound). */
+      if (event instanceof ExchangeSendingEvent) {
+        ExchangeSendingEvent ese = (ExchangeSendingEvent) event;
+        SpanDecorator sd = CamelTracer.TRACER.getSpanDecorator(ese.getEndpoint());
+        if (!sd.shouldStartNewSpan()) {
+          return;
+        }
+
+        Span span =
+            CamelTracer.TRACER.startSpan(
+                sd.getOperationName(ese.getExchange(), ese.getEndpoint()),
+                sd.getInitiatorSpanKind());
+        sd.pre(span, ese.getExchange(), ese.getEndpoint(), CamelDirection.OUTBOUND);
+        CamelPropagationUtil.injectParent(
+            Context.current(), ese.getExchange().getIn().getHeaders());
+        ActiveSpanManager.activate(ese.getExchange(), span);
+
+        if (LOG.isTraceEnabled()) {
+          LOG.trace("[Exchange sending] Initiator span started " + span);
+        }
+        /** Camel finished sending (outbound). Finish span and remove it from CAMEL holder. */
+      } else if (event instanceof ExchangeSentEvent) {
+        ExchangeSentEvent ese = (ExchangeSentEvent) event;
+        SpanDecorator sd = CamelTracer.TRACER.getSpanDecorator(ese.getEndpoint());
+        if (!sd.shouldStartNewSpan()) {
+          return;
+        }
+        Span span = ActiveSpanManager.getSpan(ese.getExchange());
+        if (span != null) {
+          if (LOG.isTraceEnabled()) {
+            LOG.trace("[Exchange sent] Initiator span finished " + span);
+          }
+          sd.post(span, ese.getExchange(), ese.getEndpoint());
+          ActiveSpanManager.deactivate(ese.getExchange());
+        } else {
+          LOG.warn("Could not find managed span for exchange " + ese.getExchange());
+        }
+      }
+    } catch (Throwable t) {
+      LOG.warn("Failed to capture tracing data", t);
+    }
+  }
```

Comment:
this method is really big. What do you think about extracting methods for each of the two branches?

Reply:
Done.